File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.cpp
class Solution {
public:
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        // Fast I/O optimization
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        
        vector<int> group(n, 0);
        int component_id = 0;
        
        for (int i = 1; i < n; ++i) {
            if (nums[i] - nums[i - 1] > maxDiff) {
                component_id++;
            }
            group[i] = component_id;
        }
        
        vector<bool> ans;
        ans.reserve(queries.size());
        for (const auto& q : queries) {
            ans.push_back(group[q[0]] == group[q[1]]);
        }
        
        return ans;
    }
};

File: 3838-path-existence-queries-in-a-graph-i/path-existence-queries-in-a-graph-i.py
class Solution:
    def pathExistenceQueries(self, n: int, nums: list[int], maxDiff: int, queries: list[list[int]]) -> list[bool]:
        # Pre-allocate array for fast, contiguous memory layout
        group = [0] * n
        curr = 0
        
        # Micro-optimization: standard for-loops with local vars are fastest for mutating elements
        for i in range(1, n):
            if nums[i] - nums[i - 1] > maxDiff:
                curr += 1
            group[i] = curr
            
        # Using map with a lambda bypasses the byte-code execution loop of a list comprehension,
        # executing the loop natively in C speed.
        return list(map(lambda q: group[q[0]] == group[q[1]], queries))